### PR TITLE
Allow spectators to be kicked after the game starts.

### LINKED ISSUE #3516: In-game kicking
It should be possible to kick players (and spectators) while playing a game. Sometimes a troll doesn't reveal themselves as a troll until you have already started playing a game.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/447905-in-game-kicking?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #17656: Ability for lobby host to kick spectator in game
It would be interesting to extend the power of a lobby admin (host) in game with the ability to kick spectators.

Spectators are often the cause of a game being restarted for the following reasons : 

-performance issues
-Lag issues
-Anything related to the in game "all chat"

Sometime asking a spectator to leave is enough, but all the involved player and respectful spec would like to be able to kick an undesired spec or just an afk spec causing issues.

I don't have a particular preference to implement it but i guess a "kick" button in scoreboard would do the trick with a "confirm" chat box to be sur not to missclick  "Confirm kick on xxxxx YES NO"

### PR DESCRIPTION
**Pull request by pchote**
Fixes #17656.
Closes #3516 (as kicking players is not wanted).

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoLogic.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
  */
 #endregion
 
+using System;
```
Comment (wanderrful) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoLogic.cs]: Stupid question - is this import needed? I don't see anything referencing it directly in the changes here.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs [hunk 2]
```diff
@@ -106,13 +106,19 @@ public class IngameMenuLogic : ChromeLogic
 			{
 				var gameInfoPanel = Game.LoadWidget(world, "GAME_INFO_PANEL", panelRoot, new WidgetArgs()
 				{
-					{ "activePanel", activePanel }
+					{ "activePanel", activePanel },
+					{ "hideMenu", (Action<bool>)RequestHideMenu }
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs]: Is there a reason you defined this as void method and cast it to `Action<bool>` here? Apparently we defined such explicitly in other places: https://github.com/OpenRA/OpenRA/blob/d6436858a90ee3299f379ec827b0065d7183492f/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs#L83

**Comment by pchote** [comment]
Updated (first push) and rebased (second).

**Comment by Mailaender** [comment]
For some reason I can't kick myself. Maybe because the nickname is identical:

image

image

I can't spot an obvious problem in your code though.

**Comment by Mailaender** [comment]
Also doesn't work with different nicknames and no bots:

image


image

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 3]
```diff
@@ -573,6 +573,12 @@ static bool Kick(S server, Connection conn, Session.Client client, string s)
 				return true;
 			}
 
+			if (server.State == ServerState.GameStarted && !client.IsObserver)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: I think this is checking whether the admin is an observer. This should be using the `kickClientID` or `kickConn` or `kickClient` instead.

Comment (pchote) [reply OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: 🤦‍♂

**Comment by pchote** [comment]
Fixed.

**Review by Mailaender — DISMISSED** [review]
Works as promised across mods.

**Review by jrb0001 — CHANGES_REQUESTED** [review]
You should increase the order protocol version. This is the first command order which is valid after the game started.

**Comment by pchote** [comment]
Done.

**Comment by pchote** [comment]
Updated to fix kicking not working if the game was not force-started.

**Review by abcdefg30 — APPROVED** [review]
Feature works. Imho we should think about polishing the dialogue a bit in a follow-up (this might work together with fixing the other issues after the IPv6 PR).
grafik

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1f9c2e72d5085c7a2ea32ed9fda83920e882cb55...5b7446eadce62de7e2747d0468641c8cb052b48d)

**Pull request merged**